Add Dapr sidecar configuration investigations (#44)
kyma-project/control-plane: docs/internal/investigations/remote-kyma-runtime-configuration.md

# Remote Kyma Runtime configuration

## Introduction

There is a need for configuring Dapr sidecars on a Kyma Runtime from one central Control Plane.

The proposed solution is to create one central component that would hold configurations for all the Runtimes. The Runtime will have a new component, a Control Plane Runtime Agent, that will periodically fetch the configuration and apply it on the Kyma Runtime. However, the Dapr 
sidecar have a shortcoming. As described [here](https://github.com/dapr/dapr/issues/1172), for the configuration to be applied
the whole pod needs to restart.

## Solutions

There are a few solutions for this problem:

> **Control Plane Runtime Agent will restart every pod with a dapr sidecar**

Reviewer: ```suggestion
**Control Plane Runtime Agent will restart every Pod with a Dapr sidecar**
```

> **Control Plane Runtime Agent will restart every pod with a dapr sidecar**

Reviewer: Consider changing **bold** formatting into H3 headings (###). It would be easier to link to these solutions in the future, if needed. However, I do not have a strong opinion about that so I leave the decision to you.

> This solution may be the simplest, but it could add to many responsibilities for one component, it would fetch the configuration, apply it and then restart all the pods

Reviewer: ```suggestion
This solution may be the simplest, but it could add too many responsibilities for one component. A single Control Plane Runtime Agent would fetch the configuration, apply it and then restart all Pods.

```

Pros:

> - No new components on the Runtime, less resource consuming

Reviewer: ```suggestion
- No new components in the Runtime 
- Less resource-consuming
```

Cons:
- Too many responsibilites for one component

> **New component responsible for restarting the pods**

Reviewer: ```suggestion
**New component responsible for restarting Pods**
```

> This solution adds a new component that will be triggered by the Agent and will restart the pods mentioned in the configuration.

Reviewer: ```suggestion
This solution adds a new component that will be triggered by the Agent and will restart Pods mentioned in the configuration.

```

Pros:
- Responsibilites distributed between components

Cons:
- Another component will definitely mean more resource consumption, which we want to avoid

**Dapr contribution that would allow configuration reloading**

> This solution will surely need the most amount work. It seems like a really big change to the Dapr ecosystem and the approach for that is still being [discussed](https://github.com/dapr/dapr/issues/1172#issuecomment-610568718). We would have to create a proposal, wait for it to be accepted and then develop the feature.

Reviewer: ```suggestion
This solution will surely be the most time-consuming. It seems like a really big change to the Dapr ecosystem and the approach to that is still being [discussed](https://github.com/dapr/dapr/issues/1172#issuecomment-610568718). We would have to create a proposal, wait for it to be accepted, and then develop the feature.
```

Pros:

> - Configuration reload out of the box

Reviewer: ```suggestion
- Configuration reload out-of-the-box
```

# Remote Kyma Runtime configuration

> We need to configure Dapr sidecars on a Kyma Runtime from one central Kyma Control Plane.

Reviewer: ```suggestion
We need to configure Dapr sidecars in a Kyma Runtime from one central Kyma Control Plane.
```

The proposed solution is to create one central component that would hold configurations for all Runtimes. Each Runtime will have a new component called Control Plane Runtime Agent that will periodically fetch the configuration and apply it to a Kyma Runtime. However, the Dapr 

> sidecars have an [issue](https://github.com/dapr/dapr/issues/1172) as the whole Pod needs to restart for the configuration to be applied.

Reviewer: ```suggestion
sidecars have an [issue](https://github.com/dapr/dapr/issues/1172) concerning the fact that the whole Pod needs to restart for the configuration to be applied.
```

## Solutions

There are a few solutions for this problem:

### Control Plane Runtime Agent will restart every Pod with a Dapr sidecar

> This solution may be the simplest, but it could add too many responsibilities for one component. A single Control Plane Runtime Agent would fetch the configuration, apply it and then restart all Pods.

Reviewer: ```suggestion
This solution may be the simplest, but it could add too many responsibilities for one component. A single Control Plane Runtime Agent would fetch the configuration, apply it, and then restart all Pods.
```

Pros:
- No new components in the Runtime 
- Less resource-consuming

Cons:
- Too many responsibilites for one component

### New component responsible for restarting Pods
This solution adds a new component that will be triggered by the Agent and will restart Pods mentioned in the configuration.

Pros:
- Responsibilites distributed between components

Cons:
- Another component will definitely mean more resource consumption, which we want to avoid

### Dapr contribution that would allow configuration reloading

> This solution will surely be the most time-consuming. It seems like a really big change to the Dapr ecosystem and the approach to that is still being [discussed](https://github.com/dapr/dapr/issues/1172#issuecomment-610568718). We would have to create a proposal, wait for it to be accepted, and then develop the feature.

Reviewer: ```suggestion
This solution will surely be the most time-consuming. It seems like a really big change to the Dapr ecosystem and the approach to that is still being [discussed](https://github.com/dapr/dapr/issues/1172#issuecomment-610568718). We would have to create a proposal, wait for it to be accepted, and then develop the feature.

```
